src/test/diagnostics.test.ts

```diff
@@ -0,0 +1,219 @@
+		return {
+			file,
+			messageText: message,
+			start: 0,
+			length: 4,
+			category: ts.DiagnosticCategory.Error,
+			code: 33
+		};
+	};
+	let publishSpy: sinon.SinonSpy;
+	const sourceFile1 = ts.createSourceFile('/file1.ts', '', ts.ScriptTarget.ES2015);
+	const file1FailureA = createTSFileDiagnostic('Failure A', sourceFile1);
+	const file1FailureB = createTSFileDiagnostic('Failure B', sourceFile1);
+	const failureADiagnostic = {
+		message: 'Failure A',
+		range: { end: { character: 4, line: 0 }, start: { character: 0, line: 0 } },
+		severity: 1,
+		code: 33,
+		source: 'ts'
+	};
+	const failureBDiagnostic = {
+		message: 'Failure B',
+		range: { end: { character: 4, line: 0 }, start: { character: 0, line: 0 } },
+		severity: 1,
+		code: 33,
+		source: 'ts'
+	};
+
+	beforeEach(() => {
+		publishSpy = sinon.spy();
+		langClient = {
+			textDocumentXcontent(params: TextDocumentContentParams, childOf?: Span) {
+				return Promise.reject(new Error('not implemented')); // Promise<TextDocumentItem>;
+			},
+			workspaceXfiles(params: WorkspaceFilesParams, childOf?: Span) {
+				return Promise.reject(new Error('not implemented')); // Promise<TextDocumentIdentifier[]>
+			},
+			windowLogMessage(params: LogMessageParams) {
+				// nop
+			},
+			xcacheGet(params: CacheGetParams, childOf?: Span) {
+				return Promise.reject(new Error('not implemented')); // nop
+			},
+			xcacheSet(params: CacheSetParams) {
+				// nop
+			},
+			textDocumentPublishDiagnostics : publishSpy
+		};
+
+	});
+
+	it('should not update if there are no changes', () => {
+		diagnosticsManager = new DiagnosticsPublisher(langClient);
+		diagnosticsManager.updateFileDiagnostics([]);
+		sinon.assert.notCalled(publishSpy);
+	});
+
+	it('should translate a diagnostic correctly', () => {
+		diagnosticsManager = new DiagnosticsPublisher(langClient);
+		diagnosticsManager.updateFileDiagnostics([file1FailureA]);
+		sinon.assert.calledWithExactly(publishSpy, {
+			diagnostics: [failureADiagnostic],
+			uri: 'file:///file1.ts'
+		});
+	});
+
+	it('should group diagnostics by file', () => {
+		diagnosticsManager = new DiagnosticsPublisher(langClient);
+		diagnosticsManager.updateFileDiagnostics([file1FailureA, file1FailureB]);
+		sinon.assert.calledWithExactly(publishSpy, {
+			diagnostics: [failureADiagnostic, failureBDiagnostic],
+			uri: 'file:///file1.ts'
+		});
+	});
+
+	it('should publish empty diagnostics when file is fixed', () => {
+		diagnosticsManager = new DiagnosticsPublisher(langClient);
+		diagnosticsManager.updateFileDiagnostics([file1FailureA]);
+		sinon.assert.calledWithExactly(publishSpy, {
+			diagnostics: [failureADiagnostic],
+			uri: 'file:///file1.ts'
+		});
+		diagnosticsManager.updateFileDiagnostics([]);
+		sinon.assert.calledWithExactly(publishSpy, {
+			diagnostics: [],
+			uri: 'file:///file1.ts'
+		});
+	});
+});
+
+// Integration tests that include TS compilation.
+describe('Diagnostics', () => {
```

Comment:
Still not convinced we should have these here.

The expected behaviour of the ProjectManager is to call the DiagnosticHandler, so there should be tests in the ProjectManager suite that assert it is called. That's where all the setup/teardown code is located.

All tests are to prevent regressions